src/java.base/share/classes/jdk/internal/constant/ArrayClassDescImpl.java

```diff
@@ -0,0 +1,157 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.constant;
+
+import java.lang.constant.ClassDesc;
+import java.lang.invoke.MethodHandles;
+
+import jdk.internal.vm.annotation.Stable;
+import sun.invoke.util.Wrapper;
+
+import static java.lang.constant.ConstantDescs.CD_void;
+
+import static jdk.internal.constant.ConstantUtils.MAX_ARRAY_TYPE_DESC_DIMENSIONS;
+
+/**
+ * An array class descriptor.
+ * Restrictions: <ul>
+ * <li>{@code rank} must be in {@code [1, 255]}
+ * <li>{@code element} must not be void or array
+ * </ul>
+ */
+public final class ArrayClassDescImpl implements ClassDesc {
+    private final ClassDesc element;
+    private final int rank;
+    private @Stable String cachedDescriptorString;
+
+    public static ArrayClassDescImpl ofValidatedDescriptor(String desc) {
+        assert desc.charAt(0) == '[';
+        var lastChar = desc.charAt(desc.length() - 1);
+        ArrayClassDescImpl ret;
+        if (lastChar != ';') {
+            // Primitive element arrays
+            ret = ofValidated(Wrapper.forBasicType(lastChar).basicClassDescriptor(), desc.length() - 1);
+        } else {
+            int level = ConstantUtils.arrayDepth(desc, 0);
+            ret = ofValidated(ClassOrInterfaceDescImpl.ofValidated(desc.substring(level)), level);
+        }
+        ret.cachedDescriptorString = desc;
+        return ret;
+    }
+
+    public static ArrayClassDescImpl ofValidated(ClassDesc element, int rank) {
+        assert !element.isArray() && element != CD_void;
+        assert rank > 0 && rank <= MAX_ARRAY_TYPE_DESC_DIMENSIONS;
+
+        return new ArrayClassDescImpl(element, rank);
+    }
+
+    private ArrayClassDescImpl(ClassDesc element, int rank) {
+        this.element = element;
+        this.rank = rank;
+    }
+
+    @Override
+    public ClassDesc arrayType() {
+        if (rank == MAX_ARRAY_TYPE_DESC_DIMENSIONS)
+            throw new IllegalStateException(
+                "Cannot create an array type descriptor with more than "
+                        + MAX_ARRAY_TYPE_DESC_DIMENSIONS + " dimensions");
```

Comment:
```suggestion
        ConstantUtils.validateArrayDepth(rank + 1);
```

Reply:
Unfortunately, this has to throw ISE while `validateArrayDepth` throws IAE.

Reply:
I have extracted this to a new utility method, with a switch on the thrown exception type.